libraries/persistent_store/src/driver.rs

```diff
@@ -0,0 +1,630 @@
+        page: usize,
+
+        /// The first word in the page with a wrong value.
+        word: usize,
+
+        /// How many times the word has been written according to the store.
+        ///
+        /// This value is exact only for the metadata of the page. For the content of the page, it
+        /// is set to:
+        /// - 0 if the word is after the tail. Such word should not have been written.
+        /// - 1 if the word is before the tail. Such word may or may not have been written.
+        store: usize,
+
+        /// How many times the word has been written according to the model.
+        ///
+        /// This value is exact only for the metadata of the page. For the content of the page, it
+        /// is set to:
+        /// - 0 if the word was not written.
+        /// - 1 if the word was written.
+        model: usize,
+    },
+}
+
+impl StoreDriver {
+    /// Provides read-only access to the storage.
+    pub fn storage(&self) -> &BufferStorage {
+        match self {
+            StoreDriver::On(x) => x.store().storage(),
+            StoreDriver::Off(x) => x.storage(),
+        }
+    }
+
+    /// Extracts the power-on version of the driver.
+    pub fn on(self) -> Option<StoreDriverOn> {
+        match self {
+            StoreDriver::On(x) => Some(x),
+            StoreDriver::Off(_) => None,
+        }
+    }
+
+    /// Powers on the store if not already on.
+    pub fn power_on(self) -> Result<StoreDriverOn, StoreInvariant> {
+        match self {
+            StoreDriver::On(x) => Ok(x),
+            StoreDriver::Off(x) => x.power_on(),
+        }
+    }
+
+    /// Extracts the power-off version of the driver.
+    pub fn off(self) -> Option<StoreDriverOff> {
+        match self {
+            StoreDriver::On(_) => None,
+            StoreDriver::Off(x) => Some(x),
+        }
+    }
+}
+
+impl StoreDriverOff {
+    /// Starts a simulation with a clean storage given its configuration.
+    pub fn new(options: BufferOptions, num_pages: usize) -> StoreDriverOff {
+        let storage = vec![0xff; num_pages * options.page_size].into_boxed_slice();
+        let storage = BufferStorage::new(storage, options);
+        StoreDriverOff::new_dirty(storage)
+    }
+
+    /// Starts a simulation from an existing storage.
+    pub fn new_dirty(storage: BufferStorage) -> StoreDriverOff {
+        let format = Format::new(&storage).unwrap();
+        StoreDriverOff {
+            storage,
+            model: StoreModel::new(format),
+            complete: None,
+        }
+    }
+
+    /// Provides read-only access to the storage.
+    pub fn storage(&self) -> &BufferStorage {
+        &self.storage
+    }
+
+    /// Provides mutable access to the storage.
+    pub fn storage_mut(&mut self) -> &mut BufferStorage {
+        &mut self.storage
+    }
+
+    /// Provides read-only access to the model.
+    pub fn model(&self) -> &StoreModel {
+        &self.model
+    }
+
+    /// Powers on the store without interruption.
```

Comment:
The intended level is minimal. The code is usually clearer than the documentation would be, so I don't think it's worth documenting too much.

Those functions are public but only available in "std" environment, i.e. for testing and fuzzing. They are meant to panic as often as possible except for the store invariants above for which an error is returned, such that the fuzzing code can print additional information before crashing. There's more store invariants that could be added (like storage preconditions), so this is mostly a compromise between panicking right away, or popping up the error up to fuzzing.

Reply:
Okay, just wondered if people understand if the browse the generated API documentation, or what the usual way is to make readers understand.

Reply:
I think this code has too much chance to evolve that documenting anything in stone now is not worth it. The best to understand what a function does is to read the code and to understand why we need it to read its usage in fuzzing.

I'll probably change a bit the driver and fuzzing to only do interruptions near the end of entropy. This would make fuzzing faster and also improve coverage distribution I think (because a prefix of an artifact will not only check less things but also other things). But I'm still not completely sure and it will take some time to design and implement.